docs/articles/persistence/custom-persistence-provider.md

```diff
@@ -0,0 +1,466 @@
+
+##### Code Sample
+
+`ByPersistenceIdSql` in this example refers to this SQL query statement:
+
+[!code-csharp[ByPersistenceIdSql](../../../src/examples/Akka.Persistence.Custom/Journal/SqliteJournal.cs?name=ByPersistenceIdSql "ByPersistenceIdSql SQL Statement")]
+
+SQLite code example:
+
+[!code-csharp[ReplayMessagesAsync](../../../src/examples/Akka.Persistence.Custom/Journal/SqliteJournal.cs?name=ReplayMessagesAsync "ReplayMessagesAsync method implementation")]
+
+#### ReadHighestSequenceNrAsync
+
+```c#
+Task<long> ReadHighestSequenceNrAsync(
+    string persistenceId, 
+    long fromSequenceNr)
+```
+
+* **`persistenceId`**: Persistent actor identifier
+* **`fromSequenceNr`**: Hint where to start searching for the highest sequence number. When a persistent actor is recovering this `fromSequenceNr` will the sequence number of the used snapshot, or `0L` if no snapshot is used.
+
+This call is protected with a circuit-breaker.
+
+This method should asynchronously reads the highest stored sequence number for provided `persistenceId`. The persistent actor will use the highest sequence number after recovery as the starting point when persisting new events. This sequence number is also used as `toSequenceNr` in subsequent calls to `ReplayMessagesAsync` unless the user has specified a lower `toSequenceNr`. Journal must maintain the highest sequence number and never decrease it.
+
+Please also note that requests for the highest sequence number may be made concurrently to writes executing for the same `persistenceId`, in particular it is possible that a restarting actor tries to recover before its outstanding writes have completed.
+
+`HighestSequenceNrSql` in this example refers to this SQL query statement:
+
+[!code-csharp[HighestSequenceNrSql](../../../src/examples/Akka.Persistence.Custom/Journal/SqliteJournal.cs?name=HighestSequenceNrSql "HighestSequenceNrSql SQL Statement")]
+
+SQLite code example:
+
+[!code-csharp[ReadHighestSequenceNrAsync](../../../src/examples/Akka.Persistence.Custom/Journal/SqliteJournal.cs?name=ReadHighestSequenceNrAsync "ReadHighestSequenceNrAsync method implementation")]
+
+#### WriteMessagesAsync
+
+```c#
+protected abstract Task<IImmutableList<Exception>> WriteMessagesAsync(
+    IEnumerable<AtomicWrite> messages)
+```
+
+Asynchronously writes a batch of persistent messages to the journal.
+
+This call is protected with a circuit-breaker.
+
+Calls to this method are serialized by the enclosing journal actor. If you spawn work in asynchronous tasks it is alright that they complete the futures in any order, but the actual writes for a specific persistenceId should be serialized to avoid issues such as events of a later write are visible to consumers (query side, or replay) before the events of an earlier write are visible. A `PersistentActor` will not send a new `WriteMessages` request before the previous one has been completed.
```

Comment:
Not 100% true - `PersistAsync` explicitly does this. The journal may save events out of order, but they're ACKed back to the caller in the original order in which they were sent. Since the caller sets the `SeqNo`, which determines the ordering, this operation is safe.